datafusion/execution/src/object_store.rs

```diff
@@ -89,6 +89,138 @@ pub trait ObjectStoreProvider: Send + Sync + 'static {
+pub trait ObjectStoreManager: Send + Sync + std::fmt::Debug + 'static {
+    /// If a store with the same schema and host existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        scheme: &str,
+        host: &str,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Get a suitable store for the provided URL. For example:
+    ///
+    /// - URL with scheme `file:///` or no schema will return the default LocalFS store
+    /// - URL with scheme `s3://bucket/` will return the S3 store
+    /// - URL with scheme `hdfs://hostname:port/` will return the hdfs store
+    fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>>;
+}
+
+/// The default [`ObjectStoreManager`] is with no `ObjectStoreProvider`.
+pub struct DefaultObjectStoreManager {
+    /// A map from scheme to object store that serve list / read operations for the store
+    object_stores: DashMap<String, Arc<dyn ObjectStore>>,
+}
+
+impl std::fmt::Debug for DefaultObjectStoreManager {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("DefaultObjectStoreManager")
+            .field(
+                "schemes",
+                &self
+                    .object_stores
+                    .iter()
+                    .map(|o| o.key().clone())
+                    .collect::<Vec<_>>(),
+            )
+            .finish()
+    }
+}
+
+impl Default for DefaultObjectStoreManager {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DefaultObjectStoreManager {
+    /// This will register [`LocalFileSystem`] to handle `file://` paths
+    pub fn new() -> Self {
+        let object_stores: DashMap<String, Arc<dyn ObjectStore>> = DashMap::new();
+        object_stores.insert("file://".to_string(), Arc::new(LocalFileSystem::new()));
+        Self { object_stores }
+    }
+}
+
+impl ObjectStoreManager for DefaultObjectStoreManager {
+    fn register_store(
+        &self,
+        scheme: &str,
+        host: &str,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        let s = format!("{scheme}://{host}");
+        self.object_stores.insert(s, store)
+    }
+
+    /// Get a suitable store for the provided URL. For example:
+    ///
+    /// - URL with scheme `file:///` or no schema will return the default LocalFS store
+    /// - URL with scheme `s3://bucket/` will return the S3 store if it's registered
+    /// - URL with scheme `hdfs://hostname:port/` will return the hdfs store if it's registered
+    fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>> {
+        let s = &url[url::Position::BeforeScheme..url::Position::BeforePath];
+        self.object_stores
+            .get(s)
+            .map(|o| o.value().clone())
+            .ok_or_else(|| {
+                DataFusionError::Internal(format!(
+                    "No suitable object store found for {url}"
+                ))
+            })
+    }
+}
```

Comment:
I'm confused by this. This is the same interface as `ObjectStoreRegistry`. What is the functional difference that requires wrapping this in another layer of indirection?

Reply:
The reason for this is to make the cache transparent for the caller. An example usage is here,
https://github.com/yahoNanJing/arrow-ballista/commit/48c703b3c62235d5e70f19a1bb7a39d9584138ee

Reply:
In Ballista, we can create an `ObjectStoreRegistry` with a `CacheBasedObjectStoreManager`. Then every `ObjectStore` got by `ObjectStoreRegistry` will be wrapped with a cache layer.